jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/MemoryEndPointPipe.java

```diff
@@ -0,0 +1,344 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HexFormat;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.util.thread.Invocable;
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>Memory-based implementation of {@link EndPoint.Pipe}.</p>
+ */
+public class MemoryEndPointPipe implements EndPoint.Pipe
```

Comment:
Should the LocalConnector be changed to use this?

Reply:
It is difficult, as `LocalEndPoint` is somehow incompatible with `MemoryEndPoint`.
We did not document `LocalConnector`, which we can keep as testing facility (although in `src/main`), and `MemoryConnector` is now documented.